File: systems/input.lua
local utils_table = require("utils.table")

local M = {}

M.listeners = {}

function M.register(listener)
    if not utils_table.find(M.listeners, listener) then
        table.insert(M.listeners, listener)
    end
end

function M.unregister(listener)
    utils_table.remove_value(M.listeners, listener);
end

function M.dispatch(action_id, action)
    for _, listener in ipairs(M.listeners) do
        msg.post(listener, 'input', { action_id = action_id, action = action })
    end
end

function M.on_input(action_id, action)
    if action_id == hash("touch") or action_id == hash("mouse_button_1") or action_id == hash("mouse_move") then
        M.dispatch(action_id, action)
    end
end

return M

File: src/modules/systems/Input.ts
import { Action } from '../types/Action';
import { postMessage } from './Message';
import { ActionId } from '../types/ActionId';
import { MessageId } from '../types/MessageId';

class Input {
	private listeners = new Set<url>();

	register(listener: url) {
		this.listeners.add(listener);
	}

	unregister(listener: url) {
		this.listeners.delete(listener);
	}

	onInput(actionId: ActionId, action: Action) {
		this.listeners.forEach((listener) => {
			postMessage(listener, MessageId.Input, { actionId, action });
		});
	}
}

export const input = new Input();
